### PR TITLE
Return error hash for invalid input error

### PR DESCRIPTION
**Pull request by riiwo**
This change allows FE to highlight erronous fields without much effort and
should be generaly good improvement for UX

Bump minor version

### CODE REVIEW

File: lib/salestation/web/responses.rb [hunk 1]
```diff
@@ -35,9 +35,12 @@ class Error < Response
         attribute :debug_message, Types::Coercible::String.default('')
         attribute :context, Types::Hash.default({}.freeze)
         attribute :headers, Types::Hash.default({}.freeze)
+        attribute :errors, Types::Hash.default({}.freeze)
 
         def body
-          {message: message, debug_message: debug_message}
+          body = {message: message, debug_message: debug_message}
+          body = body.merge(errors: errors) if errors != {}
```
Comment (riiwo) [inline lib/salestation/web/responses.rb:42]: `.present?` was not defined ...

**Review by kneepdee — APPROVED** [review]

File: lib/salestation/web/responses.rb [hunk 2]
```diff
@@ -35,9 +35,12 @@ class Error < Response
         attribute :debug_message, Types::Coercible::String.default('')
         attribute :context, Types::Hash.default({}.freeze)
         attribute :headers, Types::Hash.default({}.freeze)
+        attribute :errors, Types::Hash.default({}.freeze)
```
Comment (indrekj) [inline lib/salestation/web/responses.rb:38]: can you add this only to UnprocessableEntity

**Comment by indrekj** [comment]
@urmastalimaa are you ok with this as well? Currently `message` and `debug_message` are strings, we use formatted validation errors. This is nice and readable for a human, but not so much for computer. By providing the structured object, we can easily highlight erroneous fields in the UI.

**Comment by urmastalimaa** [comment]
Seems OK for form validation, but then can we name this as `form_errors`? For other semantical errors, error codes make more sense.

**Comment by indrekj** [comment]
> Seems OK for form validation, but then can we name this as form_errors? For other semantical errors, error codes make more sense.

I think so. I'd also want to show this on 422 errors. For other errors it doesn't make sense.

**Comment by deiwin** [comment]
cc @ostap0207 & @laurglia who've previously mentioned things around this topic.

**Comment by ostap0207** [comment]
Is there an example of how these are gonna look?
It would be good to decide on the format/structure of this errors and document them as well. Also, make sure that other services (elixir) can provide the same structure.

**Review by indrekj — APPROVED** [review]

**Comment by riiwo** [comment]
> Is there an example of how these are gonna look?
> It would be good to decide on the format/structure of this errors and document them as well. Also, make sure that other services (elixir) can provide the same structure.

These will look like default dry validation messages e.g. `{:age=>["must be greater than 18"]}` (I guess same way AR errors look like)

I'll add nested hash example soon

**Comment by urmastalimaa** [comment]
Also check how Ecto error messages look like, we should have a unified structure for all.

**Comment by ostap0207** [comment]
> These will look like default dry validation messages e.g. {:age=>["must be greater than 18"]} (I guess same way AR errors look like)

Not sure how these are machine readable, these will have to be parsed as well.

Machine reable might look something like `{:age=>{:error_type => "out_of_range", :lower_bound => 18]}` (just a simple exampe, that probably does not cover all the cases).

I think this could be a broader topic and that might not be solvable within this PR. Not sure if this PR is meant as some quick fix or we have some time for discussion to get this more or less right.

We had in mind of taking this as Channels OKR for Q2, but had to postpone.

What are some other thoughts around this?

**Comment by indrekj** [comment]
> Not sure how these are machine readable, these will have to be parsed as well.

It's more machine-readable in a sense that now you know which fields caused an error. Previously you didn't have this information and had to parse the strings. But you're correct that it's not fully parseable because you don't really know what exactly is wrong with the field.

**Comment by deiwin** [comment]
> I think this could be a broader topic and that might not be solvable within this PR. Not sure if this PR is meant as some quick fix or we have some time for discussion to get this more or less right.

Prism is currently solving a very specific problem. In the scope of this PR, we should agree on what is the best we can do right now in the scope of this problem that would not cause too much rework or any show-stopping problems. We don't have to solve all of the problems & questions right away.

As far as I understand, for the current problem, only the names of problematic fields are needed to highlight those to the user in the UI. We can leave further details (e.g. what's wrong about the field) undefined as long as the proposed structure is flexible enough to accommodate them in the future.

**Comment by indrekj** [comment]
Ecto errors look somewhat similar by default:
```
changeset = User.changeset(%User{}, %{age: 0, email: "mary@example.com"})
{:error, changeset} = Repo.insert(changeset)
changeset.errors #=> [age: {"is invalid", []}, name: {"can't be blank", []}]
```

**Comment by ostap0207** [comment]
> We don't have to solve all of the problems & questions right away.

Ok, yeah if these changes are gonna be private (not documented in public REST API docs) then as long as both ruby and elixir services can output the same format, it should be OK I think.

**Comment by riiwo** [comment]
!merge

**Pull request merged**